src/chunk.cc

```diff
@@ -0,0 +1,621 @@
+        {
+          isNum = false;
+        }
+      }
+    }
+    else if(target[i] == L'\\')
+    {
+      i++;
+    }
+  }
+  return ret;
+}
+
+void
+Chunk::updateTags(const vector<wstring>& parentTags)
+{
+  if(isBlank) return;
+  unsigned int last = 0;
+  wstring result;
+  result.reserve(target.size() + (2*parentTags.size()));
+  // a rough estimate - works if most number tags are 1 digit and most new tags are 3 chars or less
+  for(unsigned int i = 0, limit = target.size(); i < limit; i++)
+  {
+    if(target[i] == L'<')
+    {
+      result += target.substr(last, i-last);
+      last = i;
+      bool isNum = true;
+      for(unsigned int j = i+1; j < limit; j++)
+      {
+        if(target[j] == L'>')
+        {
+          if(isNum)
+          {
+            unsigned int n = stoul(target.substr(last+1, j-last-1));
+            if(n != 0 && n <= parentTags.size())
+            {
+              result += parentTags[n-1];
+            }
+          }
+          else
+          {
+            result += target.substr(last, j-last+1);
+          }
+          last = j+1;
+          break;
+        }
+        if(!isdigit(target[j]))
+        {
+          isNum = false;
+        }
+      }
+    }
+    else if(target[i] == L'\\')
+    {
+      i++;
+    }
+  }
+  if(last != target.size()-1)
+  {
+    result += target.substr(last);
+  }
+  target = result;
+}
+
+void
+Chunk::output(const vector<wstring>& parentTags, FILE* out = NULL)
+{
+  if(contents.size() > 0)
+  {
+    vector<wstring> tags = getTags(parentTags);
+    for(unsigned int i = 0; i < contents.size(); i++)
+    {
+      contents[i]->output(tags, out);
+    }
+  }
+  else if(isBlank)
+  {
+    if(out == NULL)
+    {
+      cout << UtfConverter::toUtf8(target);
+    }
+    else
+    {
+      fputs_unlocked(UtfConverter::toUtf8(target).c_str(), out);
+    }
+  }
+  else
+  {
+    updateTags(parentTags);
+    if(target.size() == 0)
+    {
+    }
+    else if(out == NULL)
+    {
+      cout << UtfConverter::toUtf8(wblank);
+      cout << "^";
+      cout << UtfConverter::toUtf8(target);
+      cout << "$";
+    }
+    else
+    {
+      fputs_unlocked(UtfConverter::toUtf8(wblank).c_str(), out);
+      fputc_unlocked('^', out);
+      fputs_unlocked(UtfConverter::toUtf8(target).c_str(), out);
+      fputc_unlocked('$', out);
+    }
+  }
+}
+
+void
+Chunk::output(FILE* out)
+{
+  vector<wstring> tags;
+  output(tags, out);
+}
+
+wstring
+Chunk::matchSurface()
+{
+  if(contents.size() == 0)
+  {
+    return source;
+  }
+  return target;
+}
+
+void
+Chunk::appendChild(Chunk* kid)
+{
+  contents.push_back(kid);
+}
+
+void
+Chunk::conjoin(Chunk* other)
+{
+  unsigned int lemq_loc = 0;
+  for(; lemq_loc < target.size(); lemq_loc++)
+  {
+    if(target[lemq_loc] == L'\\')
+    {
+      lemq_loc++;
+      continue;
+    }
+    else if(target[lemq_loc] == L'#')
+    {
+      break;
+    }
+  }
+  target.insert(lemq_loc, L"+" + other->target);
+  wblank = combineWblanks(other->wblank, wblank);
```

Comment:
Seems like maybe this should be the other order?

Reply:
as in `wblank, other->wblank`

Reply:
Yeah it seems like it but apparently other is the first one in order and the current chunk is the second one.

Reply:
See the conjoined LU tests

Reply:
weird, but ok